zrml/prediction-markets/src/lib.rs

```diff
@@ -2269,18 +2312,65 @@ mod pallet {
             Ok(())
         }
 
+        pub(crate) fn do_sell_complete_set(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure!(amount != BalanceOf::<T>::zero(), Error::<T>::ZeroAmount);
+
+            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+            ensure!(
+                matches!(market.scoring_rule, ScoringRule::CPMM | ScoringRule::Lmsr),
+                Error::<T>::InvalidScoringRule
+            );
+            Self::ensure_market_is_active(&market)?;
+
+            let market_account = <zrml_market_commons::Pallet<T>>::market_account(market_id);
+            ensure!(
+                T::AssetManager::free_balance(market.base_asset, &market_account) >= amount,
+                "Market account does not have sufficient reserves.",
+            );
+
+            let assets = Self::outcome_assets(market_id, &market);
+
+            // verify first.
+            for asset in assets.iter() {
+                // Ensures that the sender has sufficient amount of each
+                // share in the set.
+                ensure!(
+                    T::AssetManager::free_balance(*asset, &who) >= amount,
+                    Error::<T>::InsufficientShareBalance,
+                );
+            }
+
+            // write last.
+            for asset in assets.iter() {
+                T::AssetManager::slash(*asset, &who, amount);
+            }
+
+            T::AssetManager::transfer(market.base_asset, &market_account, &who, amount)?;
```

Comment:
Should the transfer fail but the slash succeed, the user would lose their assets but not be reimbursed in `base_asset` as the state is not reverted (at least it is not required by the function)

Reply:
Ah, yes! I've added `#[require_transactional]`. I believe the `transfer` can't fail thanks to an `ensure!` earlier in the function, but I'm a fan of just making everything transactional.

Reply:
dfd80ea468fad5c3e7bbfe701ca8719815045533